Add type annotations for mypy - Part 1

maximilianriemensberger opened the pull request:
Issue #10 

Add type annotations to all pypck files except module.py and connection.py. mypy --strict is warning/error-free on all files except those two.  Add an additonal stage for mypy (not yet strict). I intend to switch it to strict once all files have complete type annotations.

The two missing files turn out to be more work than expected since they are pretty dependent on each other. For example, full type annotations in module.py require an additional import of connection.py, which already imports module.py. It might be a good idea regardless of type annotations to break this circular dependency. So I did split those two off from this PR for now.

@alengwenus  Please let me know what you think. Thanks!

alengwenus commented:
Thank you for contributing. I really appreciate that!

Just a few comments to get this PR through:
- I use `black` as code formatter. To prevent commits being poluted by format changes, please run `black` before committing.
- Circular dependencies introduced by mypy are usually resolved by using the `TYPE_CHECKING` constants (https://docs.python.org/3/library/typing.html#typing.TYPE_CHECKING)

For convenience I added a [pre-commit](https://pre-commit.com/) hook to the `dev` branch which runs `black` on every commit. Maybe you should rebase your PR.

maximilianriemensberger commented:
Thanks for reviewing.

> * I use `black` as code formatter. To prevent commits being poluted by format changes, please run `black` before committing.

Sure. I updated the commits accordingly.

> * Circular dependencies introduced by mypy are usually resolved by using the `TYPE_CHECKING` constants (https://docs.python.org/3/library/typing.html#typing.TYPE_CHECKING)

Thanks for the hint. I'll try it out. Long-term in might be worthwhile to break the circular dependency.

> For convenience I added a [pre-commit](https://pre-commit.com/) hook to the `dev` branch which runs `black` on every commit. Maybe you should rebase your PR.

Done. Thanks for the hint. Maybe it would also make sense to have a black stage that checks but does not fix in the tox setup so that running tox runs all checks at once.

alengwenus reviewed (approved):
Nice! Thanks @maximilianriemensberger!
There are several pylint-disabled statements left. We should fix the invalid variable names later on.
